File: app/src/main/java/com/osrsd/cache/provider/InvProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.Definition;
import com.osrsd.cache.def.InvDefinition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;

@Slf4j
public final class InvProvider {

    public static Definition decode(ByteBuffer buffer, InvDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 2) {
                definition.setSize(buffer.getShort() & 0xffff);
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/InvProvider.kt
package com.osedit.cache.provider

import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.InvDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object InvProvider : Provider<InvDefinition> {

    private val log: Logger = LoggerFactory.getLogger(InvProvider::class.java)

    override fun decode(buffer: ByteBuffer, definition: InvDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                2 -> definition.size = buffer.short.toInt() and 0xffff
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }
}
